test/cpp/api/dataloader.cpp

```diff
@@ -1314,67 +1377,202 @@ class DummyChunkReader
+
+  BatchType read_chunk(size_t chunk_index) override {
+    return {};
   }
-  ASSERT_EQ(count, 10);
+
+  size_t get_chunk_count() override {
+    return chunk_count_;
+  };
+
+  void reset() override{};
+
+  const static size_t chunk_count_ = 3;
+};
+
+TEST(DataLoaderTest, ChunkDataSetEmptyBatchThrowException) {
+  const size_t prefetch_count = 1;
+  const size_t batch_size = 5;
+  DummyEmptyChunkDataReader data_reader;
+  samplers::SequentialSampler sampler(0);
+
+  datasets::SharedBatchDataset<datasets::ChunkDataSet<
+      DummyEmptyChunkDataReader,
+      samplers::SequentialSampler,
+      samplers::SequentialSampler>>
+      dataset = datasets::make_shared_dataset<datasets::ChunkDataSet<
+          DummyEmptyChunkDataReader,
+          samplers::SequentialSampler,
+          samplers::SequentialSampler>>(
+          data_reader, sampler, sampler, prefetch_count, batch_size);
+
+  auto data_loader = torch::data::make_data_loader(
+      dataset, DataLoaderOptions(batch_size).workers(0));
+
+  ASSERT_THROWS_WITH(*(data_loader->begin()), "Chunk with index 0 is empty");
 }
 
-TEST(DataTest, DataLoaderWithChunkSupportMultiWorker) {
-  const size_t kBatchSize = 13;
-  const size_t kNumChunks = 10;
+TEST(DataLoaderTest, ChunkDataSetEmptyBatchIgnoreEmptyChunk) {
+  const size_t prefetch_count = 1;
+  const size_t batch_size = 5;
+  DummyEmptyChunkDataReader data_reader;
+  samplers::SequentialSampler sampler(0);
 
-  auto dataset =
-      torch::data::datasets::make_shared_dataset<datasets::ChunkDataSet<
-          DummyChunkReader,
+  datasets::SharedBatchDataset<datasets::ChunkDataSet<
+      DummyEmptyChunkDataReader,
+      samplers::SequentialSampler,
+      samplers::SequentialSampler>>
+      dataset = datasets::make_shared_dataset<datasets::ChunkDataSet<
+          DummyEmptyChunkDataReader,
           samplers::SequentialSampler,
           samplers::SequentialSampler>>(
-          DummyChunkReader(kNumChunks, kBatchSize),
-          samplers::SequentialSampler(kNumChunks),
-          samplers::SequentialSampler(kBatchSize),
-          kBatchSize)
-          .map(transforms::BatchLambda<std::vector<int>, int>(
-              [](const std::vector<int>& x) {
-                return std::accumulate(x.begin(), x.end(), 0);
-              }));
-  auto data_loader =
-      torch::data::make_data_loader(dataset, DataLoaderOptions(kBatchSize));
-
-  int count = 0;
-  int result_sum = 0;
-  int expected_sum = 0;
-  for (int sum : *data_loader) {
-    result_sum += sum;
-    for (int i = 0; i < kBatchSize; ++i) {
-      expected_sum += count * kBatchSize + i;
+          data_reader, sampler, sampler, prefetch_count, batch_size, true);
+
+  auto data_loader = torch::data::make_data_loader(
+      dataset, DataLoaderOptions(batch_size).workers(0));
+
+  for (auto iterator = data_loader->begin(); iterator!= data_loader->end();  ++iterator) {
+    ASSERT_EQ(iterator->size(), 0);
+  }
+}
+
+struct DummyChunkDataReader2
```

Comment:
Let's rename this class. 2 is not a good name.

Reply:
Ha, I am not happy with 2 too, but I am running out of good names. Do you have any good suggestions?